util/dvsim/Scheduler.py

```diff
@@ -12,188 +12,58 @@
-            # cancelled or has run and failed. Give up on this item too.
-            if has_failed_dep:
-                self._killed.add(next_item)
-                self.item_to_status[next_item] = 'K'
-                continue
-
-            to_dispatch.append(next_item)
-
-        if not to_dispatch:
-            return
-
-        log.log(VERBOSE, "[%s]: [%s]: [dispatch]:\n%s", hms, self.name,
-                ", ".join(item.full_name for item in to_dispatch))
-
-        for item in to_dispatch:
-            self._running.add(item)
-            self.item_to_status[item] = 'D'
-            try:
-                item.launcher.launch()
-            except LauncherError as err:
-                log.error('{}'.format(err))
-                self._kill_item(item)
-
-    def _kill(self):
-        '''Kill any running items and cancel any that are waiting'''
+class Scheduler:
+    '''An object that runs one or more Deploy items'''
 
-        # Cancel any waiting items. We take a copy of self._queued to avoid
-        # iterating over the set as we modify it.
-        for item in [item for item in self._queued]:
-            self._cancel(item)
+    # Max jobs running at one time
+    max_parallel = 16
 
-        # Kill any running items. Again, take a copy of the set to avoid
-        # modifying it while iterating over it.
-        for item in [item for item in self._running]:
-            self._kill_item(item)
-
-    def _cancel(self, item):
-        '''Cancel an item that is currently queued'''
-        assert item in self._queued
-        self._queued.remove(item)
-        self._killed.add(item)
-        self.item_to_status[item] = 'K'
+    def __init__(self, items):
+        # 'scheduled' is a dictionary of Deploy objects associated with the
+        # target name, followed by the cfg name. As items in _scheduled are
+        # ready to be run (once their dependencies pass), they are moved to the
+        # _queued list, where they wait until slots are available for them to be
+        # dispatched.
+        self._scheduled = self._schedule_items(items)
+
+        # Sets of items, split up by their current state. The sets are
+        # disjoint and their union equals the keys of self.item_to_status.
+        # _queued is a list so that we dispatch things in order (relevant
+        # for things like tests where we have ordered things cleverly to
+        # try to see failures early). They are maintained for each target.
+        self._queued = {}
+        self._running = {}
+        self._passed = {}
+        self._failed = {}
+        self._killed = {}
+        for target in self._scheduled:
+            self._queued[target] = []
+            self._running[target] = set()
+            self._passed[target] = set()
+            self._failed[target] = set()
+            self._killed[target] = set()
+
+        # Maintain completion status of each target.
+        self._completed = set()
```

Comment:
As I understand it, a target gets added to `self._completed` once all its items are done. This set is then used to filter out targets from `self._scheduled`. Would it be easier to just remove the target from `self._scheduled` and get rid of `self._completed` entirely?